### PR TITLE
Radiation transport example

### PR DESCRIPTION
**Pull request by melo-gonzo**
# PhysicsNeMo Pull Request

## Radiation Transport Surrogate Model with Transolver

This PR adds a new PhysicsNeMo example under `examples/nuclear_engineering/radiation_transport/` that trains a Transolver surrogate for the 2-D linear radiation transport equation on two benchmark problems relevant for nuclear reactor assembly design and inertial confinement fusion — the **Lattice** and **Hohlraum** benchmarks from [Kusch et al. 2025](https://arxiv.org/pdf/2505.17284), with data generated by the [KiT-RT simulation code](https://github.com/KiT-RT). The example is built end-to-end on PhysicsNeMo's: `Mesh` datapipes, `DataLoader` / `Compose` / `Normalize` transforms, the `Transolver` model, the `CombinedOptimizer` (Muon + AdamW), and `physicsnemo.utils.checkpoint`. It supports distributed training, well-defined quantities of interest, a differentiable physics loss, and reusable modules for extending to other archetectures.

## Why

As with many other scientific and engineering pipelines, running simulations is the bottleneck. This workflow demonstrates how PhysicsNeMo, and models traditionally used in CFD, CAE, and other domains, can be reused for radiation transport. Because the KiT-RT code is "benchmark-style," it is a natural interface for validating scientific ML and surrogate models.

### New example tree

```text
examples/nuclear_engineering/radiation_transport/
├── README.md                 # walkthrough: science, install, dataset, training, eval
├── DATASET_CARD.md           # dataset card describing the .pmsh layout
└── src/                      # 12 Python modules, 8 YAML configs
    ├── train.py              # Hydra entry — composes case/data/model/train
    ├── inference.py          # Hydra-driven evaluation; writes metrics + figures
    ├── trainer.py            # training loop (DDP, AMP, gradient accumulation, warmup+cosine)
    ├── dataset.py            # `RTEBaseDataset` over a directory of `.pmsh/` stores
    ├── loader.py             # `TransolverAdapter`, `collate_no_padding`, `build_dataloaders`
    ├── transforms.py         # RTE-specific `Transform`s registered with the datapipes registry
    ├── losses.py             # region-weighted MSE + QoI physics loss
    ├── qoi.py                # differentiable QoI evaluators (final-time, T=1)
    ├── evaluation_metrics.py # field + QoI aggregators
    ├── checkpointing.py      # `best_model/` checkpointing, Muon + AdamW combo optimizer
    ├── compute_normalizations.py  # standalone CLI to produce flux / material stats YAMLs
    ├── viz.py                # 3-panel flux plot + per-region QoI scatter
    └── conf/                 # Hydra groups: case/, data/, model/, train/, inference/
```

### Data layout

Each simulation is one `<name>.pmsh/` directory (written by `physicsnemo.mesh.Mesh.save`) next to a `<name>.attrs.json` sidecar. `RTEBaseDataset._load` uses `physicsnemo.mesh.Mesh.load` for the memmap tensors and reads `raw_attrs` from the sidecar, exposing it as a `NonTensorData` `metadata` entry on the returned `TensorDict`. Splits are basename arrays; the reader appends `.pmsh` when opening stores.

### Training

- DDP-ready via `torchrun --nproc_per_node=N src/train.py`; single-process
  works via plain `python` (no DDP-specific code is gated on launch
  detection beyond what `DistributedManager` provides).
- AMP via `torch.amp.autocast` + `GradScaler` (fp16) / direct autocast (bf16).
- Optimizer is Adam by default; `train.optimizer.type=muon` returns a
  `CombinedOptimizer` with `torch.optim.Muon` for 2-D weight matrices
  + AdamW for everything else.
- Scheduler is `SequentialLR([LinearLR, CosineAnnealingLR])` for warmup
  + cosine annealing.
- Single best-by-val_loss checkpoint kept at `checkpoints/best_model/`.

## Known Limitations

- **Hohlraum boundary input flux** is present in the source data but is
  not used as a model input in this example.

## Checklist

- [x] I am familiar with the [Contributing Guidelines](https://github.com/NVIDIA/physicsnemo/blob/main/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.
- [x] The [CHANGELOG.md](https://github.com/NVIDIA/physicsnemo/blob/main/CHANGELOG.md) is up to date with these changes.
- [ ] An [issue](https://github.com/NVIDIA/physicsnemo/issues) is linked to this pull request.
- [ ] If I am implementing a new model or modifying any existing model, I have followed the [Models Implementation Coding Standards](https://github.com/NVIDIA/physicsnemo/wiki/Coding-standards-for-models-implementation).

## Review Process

**All PRs are reviewed by the PhysicsNeMo team before merging.**

Depending on which files are changed, GitHub may automatically assign a maintainer for review.

We are also testing AI-based code review tools (e.g., Greptile), which may add automated comments with a confidence score.
This score reflects the AI’s assessment of merge readiness and is **not** a qualitative judgment of your work, nor is
it an indication that the PR will be accepted / rejected.

AI-generated feedback should be reviewed critically for usefulness.
You are not required to respond to every AI comment, but they are intended to help both authors and reviewers.
Please react to Greptile comments with 👍 or 👎 to provide feedback on their accuracy.

### CODE REVIEW

File: examples/nuclear_engineering/radiation_transport/src/trainer.py [hunk 1]
```diff
@@ -0,0 +1,689 @@
...
+)
+
+
+def set_seed(seed: int) -> None:
+    """Set random seed for reproducibility across all RNGs."""
+    random.seed(seed)
+    np.random.seed(seed)
+    torch.manual_seed(seed)
+    torch.cuda.manual_seed_all(seed)
+    torch.backends.cudnn.benchmark = True
```
[the remark this answers is not in the document]
Comment (melo-gonzo) [reply examples/nuclear_engineering/radiation_transport/src/trainer.py]: Addressed (dropped line).[9d509f1](https://github.com/NVIDIA/physicsnemo/pull/1647/commits/9d509f1e8930474f16eb03ef159f779e0071a870)

File: examples/nuclear_engineering/radiation_transport/src/trainer.py [hunk 2]
```diff
@@ -0,0 +1,689 @@
...
+
+    Reads ``cfg.case.type``, ``cfg.train.amp*``, and
+    ``cfg.train.gradient_accumulation_steps`` directly so callers only
+    thread the per-epoch ``loss_cfg`` (which is pre-processed by
+    :func:`losses.parse_loss_config` and varies per epoch via warmup).
+    """
+    case_type = cfg.case.type
+    use_amp, amp_dtype = _parse_amp(cfg)
+    accum_steps = cfg.train.get("gradient_accumulation_steps", 1)
+    max_grad_norm = 10.0
```
[the remark this answers is not in the document]
Comment (melo-gonzo) [reply examples/nuclear_engineering/radiation_transport/src/trainer.py]: Addressed in [a164911](https://github.com/NVIDIA/physicsnemo/pull/1647/commits/a16491172b47e85774308f0201e0e5dcb95347f0)

File: examples/nuclear_engineering/radiation_transport/src/train.py [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+from checkpointing import create_optimizer, resume_if_available
+from loader import build_dataloaders, collate_no_padding
+from losses import create_scheduler, parse_loss_config
+from trainer import (
+    _parse_amp,
+    run_training_loop,
+    set_seed,
+    setup_training_environment,
+    wrap_ddp,
+)
```
[the remark this answers is not in the document]
Comment (melo-gonzo) [reply examples/nuclear_engineering/radiation_transport/src/train.py:41]: Update in [e0b2077](https://github.com/NVIDIA/physicsnemo/pull/1647/commits/e0b2077c05546775672fe98cef0d372280cbbeae)

File: examples/nuclear_engineering/radiation_transport/src/conf/train/base.yaml [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+
+learning_rate: 3.0e-5
+min_learning_rate: 1.0e-6
+warmup_epochs: 10
+max_grad_norm: 10.0                                 # gradient L2-norm clip applied each optimizer step
+
+pretrain_checkpoint: null
+resume_checkpoint: null
+
+# objective = mse_weight * regression_mse + physics_loss.weight * (regression_mse + qoi_loss); region_weighted swaps regression_mse for the weighted variant.
```
Comment (mnabian) [inline examples/nuclear_engineering/radiation_transport/src/conf/train/base.yaml]: The comment claims `objective = mse_weight * regression_mse + physics_loss.weight * (regression_mse + qoi_loss)`, but `trainer.compute_losses` (line 286 in `src/trainer.py`) actually computes `loss = mse_w * loss_mse + physics_w * loss_qoi` — i.e., `physics_loss.weight` multiplies only `qoi_loss`, not `regression_mse + qoi_loss`. Consider updating the comment.

File: examples/nuclear_engineering/radiation_transport/src/dataset.py [hunk 5]
```diff
@@ -0,0 +1,432 @@
...
+        return td, metadata
+
+    def _load_and_transform(self, index, stream=None):
+        """Stream-aware variant of ``_load`` used by the prefetch path.
+
+        The base class's prefetch path goes through ``self.reader[index]``
+        directly (skipping ``self._load``), so we override here too. Thread
+        pool + CUDA-stream wiring is inherited from the base class.
+        """
+        from physicsnemo.datapipes.protocols import _PrefetchResult
```
Comment (mnabian) [inline examples/nuclear_engineering/radiation_transport/src/dataset.py]: This is a private API import. Unrelated to this PR, but if this method is useful in custom datapipes, should we consider making it a public API?

File: examples/nuclear_engineering/radiation_transport/src/qoi.py [hunk 6]
```diff
@@ -0,0 +1,218 @@
...
+
+    if scalar_flux.ndim != 2:
+        raise ValueError(f"Expected scalar_flux shape (T, N), got {scalar_flux.shape}")
+
+    x = cell_centers[:, 0]
+    y = cell_centers[:, 1]
+    sigma_a = sigma_t - sigma_s
+
+    xy_corrector = -3.5
+    lbounds = torch.tensor([1.0, 2.0, 3.0, 4.0, 5.0]) + xy_corrector
```
Comment (mnabian) [inline examples/nuclear_engineering/radiation_transport/src/qoi.py]: Hot path: `evaluate_lattice_qoi_torch` is called twice per training step when physics loss is enabled (once for pred, once for target under `no_grad`). `lbounds`/`ubounds` are reallocated as CPU tensors on every call and then indexed (`lbounds[k]`) inside the 25-iteration nested loop — each `(x >= lbounds[k])` comparison implicitly pushes a 0-d CPU scalar to the GPU, so the function does ~50 tiny H2D transfers per call. Hoist these to module-level constants on the correct device (or build them with `device=cell_centers.device` once at the start of the function) and pre-flatten the loop into a vectorized mask. This is much smaller than the `.item()` issues above but compounds with them on the physics-loss path.

File: examples/nuclear_engineering/radiation_transport/src/qoi.py [hunk 7]
```diff
@@ -0,0 +1,218 @@
...
+    pos_red_left_bottom = geometry_params["llr"]
+    pos_red_left_top = geometry_params["ulr"]
+    pos_red_right_top = geometry_params["urr"]
+
+    sigma_a = sigma_t - sigma_s
+
+    in_center = (x > -0.2 + cx) & (x < 0.2 + cx) & (y > -0.4 + cy) & (y < 0.4 + cy)
+    # NOTE: matches KiT-RT's behavior of using pos_red_left_bottom for both sides
+    in_vertical = (
+        (x < pos_red_left_border) & (y > pos_red_left_bottom) & (y < pos_red_left_top)
```
Comment (mnabian) [inline examples/nuclear_engineering/radiation_transport/src/qoi.py]: The hohlraum vertical-wall predicate at `qoi.py:200-204` intentionally uses `pos_red_left_bottom` for the y-lower bound on *both* the left and right walls (only the y-upper bound differs: `pos_red_left_top` vs `pos_red_right_top`). The inline `NOTE` flags this as matching KiT-RT's actual behavior, but a future refactor could easily "symmetrize" the predicate and silently break parity with the reference solver. A small unit test that builds a synthetic mesh with known `geometry_params` and asserts the per-region absorption against hand-computed values for `evaluate_hohlraum_qoi_torch` (and similarly for `evaluate_lattice_qoi_torch`) would lock the KiT-RT-compatibility behavior in.

File: examples/nuclear_engineering/radiation_transport/README.md
Comment (mnabian) [inline examples/nuclear_engineering/radiation_transport/README.md:1]: It would be really nice to add an image for what problem is being solved and how the predictions would look like.

File: examples/nuclear_engineering/radiation_transport/README.md [hunk 8]
```diff
@@ -0,0 +1,489 @@
...
+
+---
+
+## 3. Dataset
+
+### 3.1 Data source
+
+The dataset is available on Hugging Face:
+[Linear Radiation Transport][hf-rte]. Alternatively, raw simulation data
+may be curated from the [KiT-RT repositories](https://github.com/KiT-RT).
```
Comment (mnabian) [inline examples/nuclear_engineering/radiation_transport/README.md:105]: I could not find any information about how this curation should be done.

**Review by mnabian — APPROVED** [review]
LGTM! Left a few minor comments.

File: examples/nuclear_engineering/radiation_transport/src/conf/train/base.yaml (hunk 4 above)
Comment (melo-gonzo) [reply examples/nuclear_engineering/radiation_transport/src/conf/train/base.yaml]: Fixed in [bb5d5d8](https://github.com/NVIDIA/physicsnemo/pull/1647/commits/bb5d5d892f8ac442c157551eaf930fb52697caf1)

File: examples/nuclear_engineering/radiation_transport/README.md [hunk 9]
```diff
@@ -0,0 +1,489 @@
...
+    └── hohlraum_material_stats.yaml
+```
+
+### 3.3 What's in each mesh store
+
+Each `*.pmsh/` directory is one simulation, written by
+`physicsnemo.mesh.Mesh.save(...)`. The loader uses the first and final
+`scalar_flux` snapshots and ignores intermediate snapshots. The fields are:
+
+`Mesh.points` — `(N, 2)` float32 cell-center coordinates.
```
Comment (melo-gonzo) [inline examples/nuclear_engineering/radiation_transport/README.md]: Updated in [5bbd56e](https://github.com/NVIDIA/physicsnemo/pull/1647/commits/5bbd56e586b2b02f9333084f081eda213ce795f5). Missed this initially as a result of updating the dataset structure

File: examples/nuclear_engineering/radiation_transport/README.md
Comment (melo-gonzo) [reply examples/nuclear_engineering/radiation_transport/README.md:1]: Added some images! [73c6514](https://github.com/NVIDIA/physicsnemo/pull/1647/commits/73c6514ef2c69b0be8faf97fbb1fd2f7c8f55671)

File: examples/nuclear_engineering/radiation_transport/README.md (hunk 8 above)
Comment (melo-gonzo) [reply examples/nuclear_engineering/radiation_transport/README.md:105]: Curation from KiT-RT repo will be left as an exercise for the reader at this point, scripts my be added to that repo in the future.

File: examples/nuclear_engineering/radiation_transport/src/dataset.py (hunk 5 above)
Comment (melo-gonzo) [reply examples/nuclear_engineering/radiation_transport/src/dataset.py]: This is a good catch, and is the result of packing in some non-tensor fields into the custom dataset. This can be resolved by updating the dataset.

Comment (melo-gonzo) [reply examples/nuclear_engineering/radiation_transport/src/dataset.py]: I went ahead and updated this - would have been a maintenance burden and confusing for anyone reading the code. The HuggingFace dataset has also been updated to make wiring in the dataset keys smoother. [23c889e](https://github.com/NVIDIA/physicsnemo/pull/1647/commits/23c889e3c2a0b2c4c54a322c06bf67289c4d6c14)

**Comment by loliverhennigh** [comment]
Wait, I think the split file might not actually be used here.

`RTEBaseDataset` reads the split into `self.filenames`, but loading still goes through the reader by integer index, and the reader has its own sorted list of all `.pmsh` files. So unless I’m missing something, the split file controls the length of the dataset, but not which files get loaded.

Wouldn’t that mean train/val/test can end up reading overlapping files depending on the directory ordering?

**Review by loliverhennigh — APPROVED** [review]
LGTM

**Comment by melo-gonzo** [comment]
> Wait, I think the split file might not actually be used here.

Great catch! Thanks for the attention to detail, fixed in [8c9c3e4](https://github.com/NVIDIA/physicsnemo/pull/1647/commits/8c9c3e4633e6e43ac1b41cdc54ae4a891ddcebb9).

**Comment by melo-gonzo** [comment]
/ok to test [cbb1ddc](https://github.com/NVIDIA/physicsnemo/pull/1647/commits/cbb1ddc074add31421d4b720a152da1f47eb0184)

**Comment by melo-gonzo** [comment]
/ok to test [cbb1ddc](https://github.com/NVIDIA/physicsnemo/pull/1647/commits/cbb1ddc074add31421d4b720a152da1f47eb0184)

**Comment by melo-gonzo** [comment]
/ok to test [37e1413](https://github.com/NVIDIA/physicsnemo/pull/1647/commits/37e141350e76efb9eeecd2b03999280a5719a737)

**Comment by melo-gonzo** [comment]
/ok to test [25ab8c7](https://github.com/NVIDIA/physicsnemo/pull/1647/commits/25ab8c7349ca4f78c6fedf91c0522b6cd585553b)
